Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/rolling_updater.go

```diff
@@ -93,27 +78,400 @@ const (
+	desired, err := strconv.Atoi(newRc.Annotations[desiredReplicasAnnotation])
+	if err != nil {
+		return fmt.Errorf("Unable to parse annotation for %s: %s=%s",
+			newRc.Name, desiredReplicasAnnotation, newRc.Annotations[desiredReplicasAnnotation])
+	}
+
+	// This is how much per interval we can scale up or down.
+	maxIncrement := 1
+	// This is the minimum number of replicas (total of old/new) which must
+	// remain ready for the duration of the update.
+	minReady := oldRc.Spec.Replicas - maxIncrement
+	// If false, direction is up/down. If true, down/up (in-place).
+	skipFirstUp := false
+	if config.UpdatePercent != nil {
+		// Compute the scale increment based on a percentage of the new desired
+		// count.
+		maxIncrement = int(math.Ceil(float64(desired) * (math.Abs(float64(*config.UpdatePercent)) / 100)))
+		// Compute the minimum ready requirement as a percentage of the old
+		// replica count.
+		percentOld := int(math.Ceil(float64(oldRc.Spec.Replicas) * (math.Abs(float64(*config.UpdatePercent)) / 100)))
+		minReady = oldRc.Spec.Replicas - percentOld
+		// A negative percentage indicates our scale direction should be
+		// down/up instead of the default up/down.
+		if *config.UpdatePercent < 0 {
+			skipFirstUp = true
+		}
+	}
+	// Impose a floor of 1 for the minimum ready unless this is a 100% up/down
+	// scaling operation. A down/up can't have this floor because it would be
+	// impossible to maintain a minimum ready of 1 with a maximum overall pod
+	// count of 2 when scaling down first.
+	if config.UpdatePercent == nil || (*config.UpdatePercent > 0) {
+		minReady = int(math.Max(float64(1), float64(minReady)))
+	}
+	// Helpful output about what we're about to do.
+	direction := "up"
+	if skipFirstUp {
+		direction = "down"
+	}
+	fmt.Fprintf(out, "Scaling up %s from %d to %d, scaling down %s from %d to 0 (scale %s first by %d each interval, maintain at least %d ready)\n",
+		newRc.Name, newRc.Spec.Replicas, desired, oldRc.Name, oldRc.Spec.Replicas, direction, maxIncrement, minReady)
+
+	// Scale newRc and oldRc until newRc has the desired number of replicas and
+	// oldRc has 0 replicas.
+	increment := maxIncrement
+	for newRc.Spec.Replicas != desired || oldRc.Spec.Replicas != 0 {
+		if !skipFirstUp {
+			// Scale up/down. Initially this means scale up by the max increment.
+			scaledRc, err := r.scaleUp(newRc, oldRc, desired, increment, config)
+			if err != nil {
+				return err
+			}
+			newRc = scaledRc
+			skipFirstUp = true
+		} // Otherwise, scale down/up.
+
+		// Scale down as much as possible while maintaining the minimum ready
+		// amount. The scale-down informs us as to how much we can safely scale up
+		// next interval.
+		scaledRc, newIncrement, err := r.scaleDown(newRc, oldRc, desired, maxIncrement, minReady, config)
+		if err != nil {
+			return err
+		}
+		oldRc = scaledRc
+		increment = newIncrement
+
+		// Wait between down/up.
+		time.Sleep(config.UpdatePeriod)
```

Comment:
The original point of the sleeps was to give things a chance to stabilize
elsewhere in the system.  Such that a caller can add an arbitrary wait.
Also, you don't want to hotloop.  So you need some min wait period, and
then you also need the cooldown period.

On Tue, Jul 28, 2015 at 9:47 AM, Dan Mace <notifications@github.com> wrote:

> In
> Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/rolling_updater.go
> <https://github.com/openshift/origin/pull/3908#discussion_r35648372>:
>
> > +			newRc = scaledRc
> > +			skipFirstUp = true
> > +		} // Otherwise, scale down/up.
> > +
> > +		// Scale down as much as possible while maintaining the minimum ready
> > +		// amount. The scale-down informs us as to how much we can safely scale up
> > +		// next interval.
> > +		scaledRc, newIncrement, err := r.scaleDown(newRc, oldRc, desired, maxIncrement, minReady, config)
> > +		if err != nil {
> > +			return err
> > +		}
> > +		oldRc = scaledRc
> > +		increment = newIncrement
> > +
> > +		// Wait between down/up.
> > +		time.Sleep(config.UpdatePeriod)
>
> Good point. If we make efficient progress when possible and wait during
> scales, I'm not sure what these sleeps really accomplish anymore.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3908/files#r35648372>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift